src/dlls/dbgshim/dbgshim.cpp

```diff
@@ -394,13 +394,34 @@ class RuntimeStartupHelper
 
         while (numTries < 25)
         {
+            hr = EnumerateCLRs(m_processId, ppHandleArray, ppStringArray, pdwArrayLength);
+
             // EnumerateCLRs uses the OS API CreateToolhelp32Snapshot which can return ERROR_BAD_LENGTH or 
             // ERROR_PARTIAL_COPY. If we get either of those, we try wait 1/10th of a second try again (that
-            // is the recommendation of the OS API owners)
-            hr = EnumerateCLRs(m_processId, ppHandleArray, ppStringArray, pdwArrayLength);
+            // is the recommendation of the OS API owners).
             if ((hr != HRESULT_FROM_WIN32(ERROR_PARTIAL_COPY)) && (hr != HRESULT_FROM_WIN32(ERROR_BAD_LENGTH)))
             {
-                break;
+                // Just return any other error or if no handles were found (which means the coreclr module wasn't found yet).
+                if (FAILED(hr) || *ppHandleArray == NULL || *pdwArrayLength <= 0)
+                {
+                    break;
+                }
+                // If EnumerateCLRs succeeded but any of the handles are null, then sleep and retry also. This fixes a race
+                // condition where dbgshim catches the coreclr module just being loaded but before g_hContinueStartupEvent 
+                // has been initialized.
+                for (int i = 0; i < (int)*pdwArrayLength; i++)
+                {
+                    HANDLE h = (*ppHandleArray)[i];
+                    if (h != NULL && h != INVALID_HANDLE_VALUE)
+                    {
+                        break;
```

Comment:
I think you intending to break out of the while loop? This is going to break out of the for loop only. This version looks like it is looping 25 times and then giving S_OK assuming it was successful on the last iteration?

I think you also wanted to retry if any handle was NULL, but if the break applied to the outer while loop this code retries only if all handles are NULL.